foxai/explainer/fullgradeints.py

```diff
@@ -0,0 +1,462 @@
+                )
+
+    def _extract_layer_bias(self, module) -> Optional[torch.Tensor]:
+        """Extract bias of each layer
+
+        For batchnorm, the overall "bias" is different from batchnorm bias parameter.
+        Let m -> running mean, s -> running std
+        Let w -> BN weights, b -> BN bias
+        Then, ((x - m)/s)*w + b = x*w/s + (- m*w/s + b)
+        Thus (-m*w/s + b) is the effective bias of batchnorm
+
+        Args:
+            module: layer module
+
+        Returns:
+            bias if present in module else None
+        """
+
+        if isinstance(module, torch.nn.BatchNorm2d):
+            if module.running_var is None or module.running_mean is None:
+                return None
+            b = (
+                -(
+                    module.running_mean
+                    * module.weight
+                    / torch.sqrt(module.running_var + module.eps)
+                )
+                + module.bias
+            )
+            return b.data
+        elif module.bias is None:
+            return None
+        else:
+            return module.bias.data
+
+    @property
+    def biases(self) -> List[torch.Tensor]:
+        """Get biases."""
+        return self._biases
+
+    def _extract_layer_grads(  # pylint: disable = unused-argument
+        self, module, in_grad: _grad_t, out_grad: _grad_t
+    ) -> None:
+        """Collect gradient outputs from each layer, that contains bias term.
+
+        Args:
+            in_grad: input gradients
+            out_grad: output gradients
+        """
+
+        if not module.bias is None:
+            self.feature_grads.append(out_grad[0])
+
+    def get_feature_grads(
+        self, x: torch.Tensor, output_scalar: torch.Tensor
+    ) -> Tuple[torch.Tensor, List[torch.Tensor]]:
+        """Get gradients for the input and for each model layer w.r.t. input.
+
+        Args:
+            x: the input tensor
+            output_scalar: the output tensor
+
+        Returns:
+            gradients for input
+            gradients for each layer, that has bias term
+        """
+
+        # Empty feature grads list
+        self.feature_grads = []
+
+        self.model.zero_grad()
+        # Gradients w.r.t. input
+        input_gradients = torch.autograd.grad(outputs=output_scalar, inputs=x)[0]
+
+        return input_gradients, self.feature_grads
+
+
+class FullGrad:
+    """
+    Compute FullGrad saliency map and full gradient decomposition
+    """
+
+    def __init__(self, model, image_size=(3, 224, 224)):
+        self.model: torch.nn.Module = model
+        self.im_size: Tuple[int, int, int, int] = (1,) + image_size
+        self.model_ext: FullGradExtractor = FullGradExtractor(
+            model=self.model, im_size=image_size
+        )
+        self._biases: List[torch.Tensor] = self.model_ext.biases
+
+        cuda = next(self.model.parameters()).is_cuda
+        self._device = torch.device("cuda" if cuda else "cpu")
+        _ = self.check_completeness()
```

Comment:
`check_completeness` returns boolean which is discarded. We should use the result of this function or remove returning value.

Reply:
fixed